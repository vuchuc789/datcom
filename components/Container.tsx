'use client';

import styled from '@emotion/styled';
import { Box } from '@mui/material';

const StyledBox = styled(Box)`
  display: flex;
  flex-direction: column;
  height: 100vh;
`;

export const Container: React.FC<React.PropsWithChildren> = ({ children }) => {
  return <StyledBox>{children}</StyledBox>;
};
