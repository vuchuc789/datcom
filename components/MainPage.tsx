'use client';

import styled from '@emotion/styled';
import { Container, Typography } from '@mui/material';
import { useEffect } from 'react';
import useSWR from 'swr';
import { useAuth } from '../contexts/AuthContext';
import { useUtils } from '../contexts/UtilsContext';
import { ApiResponse, method } from '../types/api';
import { User } from '../types/user';

const StyledContainer = styled(Container)`
  flex-grow: 1;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

export const MainPage = () => {
  const { user } = useAuth();
  const { showAlert } = useUtils();

  const { data, error } = useSWR<User, Error>(
    () => user?.uid,
    async () => {
      const idToken = await user?.getIdToken();

      const request = new Request('/api/user', { method: method.POST });
      const headers = new Headers();
      headers.append('Authorization', `Bearer ${idToken}`);

      const res = await fetch(request, { headers });

      const parsedRes: ApiResponse<User> = await res.json();

      const data = parsedRes.data;

      if (!res.ok) {
        throw new Error(parsedRes.message || 'Đã có lỗi xảy ra !!');
      }

      if (!data) {
        throw new Error('Đã có lỗi xảy ra !!');
      }

      return data;
    },
    {
      revalidateIfStale: false,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
      shouldRetryOnError: false,
    }
  );

  useEffect(() => {
    if (error) {
      showAlert(error.message);
    }
  }, [error, showAlert]);

  return (
    <StyledContainer fixed>
      {!!data?.user?.name && (
        <Typography component="h1" variant="h4">
          {`Xin chào, ${data.user.name}`}
        </Typography>
      )}
      {!!data?.user?.uid && (
        <Typography
          component="h2"
          variant="subtitle2"
          sx={{ fontStyle: 'italic' }}
        >
          {`UID của bạn là: ${data.user.uid}`}
        </Typography>
      )}
    </StyledContainer>
  );
};
