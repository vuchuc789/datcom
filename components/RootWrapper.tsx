import { EmotionRegistry } from '../lib/EmotionRegistry';
import { ThemeProvider } from '../contexts/ThemeContext';

type RootWrapperProps = React.PropsWithChildren;

export const RootWrapper: React.FC<RootWrapperProps> = ({ children }) => {
  return (
    <EmotionRegistry>
      <ThemeProvider>{children}</ThemeProvider>
    </EmotionRegistry>
  );
};
