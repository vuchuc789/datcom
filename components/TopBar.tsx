'use client';

import styled from '@emotion/styled';
import {
  AppBar,
  Avatar,
  Box,
  Button,
  IconButton,
  Menu,
  MenuItem,
  Toolbar,
} from '@mui/material';
import Image from 'next/image';
import { useState } from 'react';
import { useAuth } from '../contexts/AuthContext';
import { useUtils } from '../contexts/UtilsContext';
import { stringAvatar, stringToColor } from '../utils/avatar';

const Logo = styled(Box)`
  flex-grow: 1;
`;
export const TopBar: React.FC = () => {
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);

  const { setLoading, showAlert } = useUtils();
  const { signInWithGooglePopup, user, signOut } = useAuth();

  return (
    <>
      <AppBar>
        <Toolbar>
          <Logo>
            <Image
              src="/rice-bowl.png"
              alt="Rice Bowl Icon"
              width={48}
              height={48}
              priority
            />
          </Logo>
          {user === null && (
            <Button
              color="inherit"
              onClick={async () => {
                try {
                  setLoading(true);
                  await signInWithGooglePopup();
                } catch (e) {
                  if (e instanceof Error) {
                    showAlert(e.message);
                  }
                } finally {
                  setLoading(false);
                }
              }}
            >
              Đăng nhập
            </Button>
          )}
          {!!user && (
            <>
              <IconButton
                id="appbar-menu-button"
                aria-controls={anchorEl ? 'appbar-menu' : undefined}
                aria-haspopup="true"
                aria-expanded={anchorEl ? 'true' : undefined}
                onClick={(e) => setAnchorEl(e.currentTarget)}
              >
                <Avatar sx={{ bgcolor: stringToColor(user.displayName || '') }}>
                  {user.photoURL ? (
                    <Image
                      src={user.photoURL}
                      alt="User Avatar"
                      width={40}
                      height={40}
                    />
                  ) : (
                    stringAvatar(user.displayName || '')
                  )}
                </Avatar>
              </IconButton>
              <Menu
                id="appbar-menu"
                anchorEl={anchorEl}
                open={!!anchorEl}
                onClose={() => {
                  setAnchorEl(null);
                }}
                MenuListProps={{
                  'aria-labelledby': 'appbar-menu-button',
                }}
                keepMounted
              >
                <MenuItem
                  onClick={async () => {
                    try {
                      await signOut();
                    } catch (e) {
                      if (e instanceof Error) {
                        showAlert(e.message);
                      }
                    } finally {
                      setAnchorEl(null);
                      setLoading(false);
                    }
                  }}
                >
                  Đăng xuất
                </MenuItem>
              </Menu>
            </>
          )}
        </Toolbar>
      </AppBar>
      <Toolbar />
    </>
  );
};
