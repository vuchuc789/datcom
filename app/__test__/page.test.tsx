import { render } from '@testing-library/react';
import RootPage from '../page';
import { RootWrapper } from '../../components/RootWrapper';

describe('Root Page', () => {
  test('render correctly', () => {
    const { container } = render(
      <RootWrapper>
        <RootPage />
      </RootWrapper>
    );

    expect(container).toMatchSnapshot();
  });
});

export {};
