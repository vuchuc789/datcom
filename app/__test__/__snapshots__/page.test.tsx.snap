// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Page render correctly 1`] = `
<div>
  <div
    class="dat-com-1ccccr5"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium dat-com-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Đăng nhập
      <span
        class="MuiTouchRipple-root dat-com-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
